src/Sylius/Bundle/CoreBundle/EventListener/CheckoutPaymentSelectionEventListener.php

```diff
@@ -0,0 +1,100 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Bundle\CoreBundle\EventListener;
+
+use Doctrine\ORM\EntityManagerInterface;
+use SM\Factory\FactoryInterface;
+use Sylius\Bundle\ResourceBundle\Controller\RedirectHandlerInterface;
+use Sylius\Bundle\ResourceBundle\Event\ResourceControllerEvent;
+use Sylius\Component\Core\Model\OrderInterface;
+use Sylius\Component\Core\Model\PaymentInterface;
+use Sylius\Component\Core\OrderCheckoutTransitions;
+use Sylius\Component\Resource\Exception\UnexpectedTypeException;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Routing\Router;
+
+/**
+ * @author Anna Walasek <anna.walasek@lakion.com>
+ */
+final class CheckoutPaymentSelectionEventListener
+{
+    /**
+     * @var FactoryInterface
+     */
+    private $stateMachineFactory;
+
+    /**
+     * @var Router
+     */
+    private $router;
+
+    /**
+     * @var EntityManagerInterface
+     */
+    private $entityManager;
+
+    /**
+     * @param FactoryInterface $stateMachineFactory
+     * @param Router $router
+     * @param EntityManagerInterface $entityManager
+     */
+    public function __construct(FactoryInterface $stateMachineFactory, Router $router, EntityManagerInterface $entityManager)
+    {
+        $this->stateMachineFactory = $stateMachineFactory;
+        $this->router = $router;
+        $this->entityManager = $entityManager;
+    }
+
+    /**
+     * @param ResourceControllerEvent $event
+     */
+    public function initializePayment(ResourceControllerEvent $event) {
+        $subject = $event->getSubject();
+
+        if (!$subject instanceof OrderInterface) {
+            throw new UnexpectedTypeException($subject, OrderInterface::class);
+        }
+
+        if (0 !== $subject->getTotal()) {
+            return;
+        }
+        
+        $stateMachine = $this->stateMachineFactory->get($subject, OrderCheckoutTransitions::GRAPH);
+
+        if ($stateMachine->can(OrderCheckoutTransitions::TRANSITION_SELECT_PAYMENT)) {
+            $stateMachine->apply(OrderCheckoutTransitions::TRANSITION_SELECT_PAYMENT);
+
+            $this->removePayments($subject);
+            
+            $url = $this->router->generate('sylius_shop_checkout_complete');
+            $event->setResponse(new RedirectResponse($url));
```

Comment:
I think you need to be a bit more specific, applying a transition and then let the "checkout resolver" to do it job? This might work, I agree, just remember to post good ideas in the first comment next time. ;)